File: src/org/jgll/grammar/Alternate.java
package org.jgll.grammar;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.List;

import org.jgll.grammar.slot.BodyGrammarSlot;
import org.jgll.grammar.slot.EpsilonGrammarSlot;
import org.jgll.grammar.slot.KeywordGrammarSlot;
import org.jgll.grammar.slot.LastGrammarSlot;
import org.jgll.grammar.slot.NonterminalGrammarSlot;

public class Alternate implements Serializable {
	
	private static final long serialVersionUID = 1L;

	private final List<BodyGrammarSlot> symbols;
	
	private final BodyGrammarSlot firstSlot;
	
	private BodyGrammarSlot condition;
	
	private final int index;
	
	public Alternate(BodyGrammarSlot firstSlot, int index) {
		
		if(firstSlot == null) {
			throw new IllegalArgumentException("firstSlot cannot be null.");
		}
		
		this.firstSlot = firstSlot;

		symbols = new ArrayList<>();
		
		this.index = index;
		
		BodyGrammarSlot current = firstSlot;
		
		if(firstSlot instanceof LastGrammarSlot) {
			symbols.add(firstSlot);
		}
		
		while(!(current instanceof LastGrammarSlot)) {
			symbols.add(current);
			current = current.next();
		}
	}
	
	public void setCondition(BodyGrammarSlot condition) {
		this.condition = condition;
	}
	
	public BodyGrammarSlot getCondition() {
		return condition;
	}
	
	public Symbol getSymbolAt(int index) {
		return symbols.get(index).getSymbol();
	}
	
	public BodyGrammarSlot getFirstSlot() {
		return firstSlot;
	}
	
	/**
	 * @return true if the alternate is of the form A ::= epsilon
	 */
	public boolean isEmpty() {
		return firstSlot instanceof EpsilonGrammarSlot;
	}
	
	public boolean isNullable() {
		if (isEmpty()) return true;
		
		BodyGrammarSlot slot = firstSlot;
		while(!(slot instanceof LastGrammarSlot)) {
			
			if(slot instanceof TerminalGrammarSlot || slot instanceof KeywordGrammarSlot)
				return false;
			
			if(slot instanceof NonterminalGrammarSlot && !slot.isNullable())
				return false;
			
			slot = slot.next();
		}
		
		return true;
	}
	
	public BodyGrammarSlot getBodyGrammarSlotAt(int index) {
		return symbols.get(index);
	}
	
	/**
	 * ::= alpha . x
	 * 
	 * @return
	 */
	public BodyGrammarSlot getLastBodySlot() {
		return symbols.get(symbols.size() - 1);
	}
	
	public int size() {
		return symbols.size();
	}
	
	public int getIndex() {
		return index;
	}
	
	public HeadGrammarSlot getNonterminalAt(int index) {
		BodyGrammarSlot bodyGrammarSlot = symbols.get(index);
		
		if(!(bodyGrammarSlot instanceof NonterminalGrammarSlot)) {
			throw new RuntimeException("The symbol at " + index + " should be a nonterminal.");
		}
		
		return ((NonterminalGrammarSlot)bodyGrammarSlot).getNonterminal();
	}
	
	public void setNonterminalAt(int index, HeadGrammarSlot head) {
		BodyGrammarSlot bodyGrammarSlot = symbols.get(index);
		
		if(!(bodyGrammarSlot instanceof NonterminalGrammarSlot)) {
			throw new RuntimeException("The symbol at " + index + " should be a nonterminal.");
		}
		
		((NonterminalGrammarSlot)bodyGrammarSlot).setNonterminal(head);
	}
	
	@Override
	public String toString() {
		StringBuilder sb = new StringBuilder();
		for(BodyGrammarSlot s : symbols) {
			sb.append(s.getSymbol()).append(" ");
		}
		return sb.toString();
	}
	
	public boolean isBinary(HeadGrammarSlot head) {
		if(! (symbols.get(0) instanceof NonterminalGrammarSlot && symbols.get(symbols.size() - 1) instanceof NonterminalGrammarSlot)) {
			return false;
		}
		
		NonterminalGrammarSlot firstNonterminal = (NonterminalGrammarSlot) symbols.get(0);
		NonterminalGrammarSlot lastNonterminal = (NonterminalGrammarSlot) symbols.get(symbols.size() - 1);
		
		return head.getNonterminal().getName().equals(firstNonterminal.getNonterminal().getNonterminal().getName()) &&
			   head.getNonterminal().getName().equals(lastNonterminal.getNonterminal().getNonterminal().getName());
	}
	
	/**
	 * 
	 * Returns true if the alternate is of the form op E.
	 * In other words, head = symbols[symobls.size - 1]
	 * 
	 */
	public boolean isUnaryPrefix(HeadGrammarSlot head) {
		
		if(isBinary(head)) {
			return false;
		}
		
		int index = symbols.size() - 1;
		if(! (symbols.get(index) instanceof NonterminalGrammarSlot)) {
			return false;
		}
		
		NonterminalGrammarSlot firstNonterminal = (NonterminalGrammarSlot) symbols.get(index);
		
		return head.getNonterminal().getName().equals(firstNonterminal.getNonterminal().getNonterminal().getName());
	}
	
	public boolean isUnaryPostfix(HeadGrammarSlot head) {
		if(isBinary(head)) {
			return false;
		}
		
		int index = 0;
		if(! (symbols.get(index) instanceof NonterminalGrammarSlot)) {
			return false;
		}
		
		NonterminalGrammarSlot lastNonterminal = (NonterminalGrammarSlot) symbols.get(index);
		
		return head.getNonterminal().getName().equals(lastNonterminal.getNonterminal().getNonterminal().getName());
	}
	
	public boolean match(List<Symbol> list) {
		
		if(list.size() != symbols.size()) {
			return false;
		}
		
		for(int i = 0; i < symbols.size(); i++) {
			if(!symbols.get(i).getSymbol().equals(list.get(i))) {
				return false;
			}
		}
		return true;
	}
	
	@Override
	public int hashCode() {
		// TODO: change it
		return 31;
	}
	
	@Override
	public boolean equals(Object obj) {
		if(this == obj) {
			return true;
		}
		
		if(!(obj instanceof Alternate)) {
			return false;
		}
		
		Alternate other = (Alternate) obj;
		
		if(this.size() != other.size()) {
			return false;
		}
		
		for(int i = 0; i < this.size(); i++) {
			BodyGrammarSlot thisSlot = symbols.get(i);
			BodyGrammarSlot otherSlot = other.symbols.get(i);
			
			if(thisSlot instanceof TerminalGrammarSlot && otherSlot instanceof TerminalGrammarSlot) {
				if(!thisSlot.getSymbol().equals(otherSlot.getSymbol())) {
					return false;
				}				
			} else if(thisSlot instanceof KeywordGrammarSlot && otherSlot instanceof KeywordGrammarSlot){
				if(!((KeywordGrammarSlot) thisSlot).getKeyword().equals(((KeywordGrammarSlot) otherSlot).getKeyword())) {
					return false;
				}
			}
			else if(thisSlot instanceof NonterminalGrammarSlot && otherSlot instanceof NonterminalGrammarSlot) {				
				Nonterminal thisNt = ((NonterminalGrammarSlot) thisSlot).getNonterminal().getNonterminal();
				Nonterminal otherNt = ((NonterminalGrammarSlot) otherSlot).getNonterminal().getNonterminal();
				
				if(thisNt == null && otherNt == null) {
					continue;
				}
				
				if(!thisNt.equals(otherNt)) {
					return false;
				}
			}
			else {
				return false;
			}
		}
		
		return true;
	}
	
}

File: src-scala/org/jgll/grammar/Alternate.scala
package org.jgll.grammar

import java.io.Serializable
import org.jgll.grammar.slot.BodyGrammarSlot
import org.jgll.grammar.slot.EpsilonGrammarSlot
import org.jgll.grammar.slot.HeadGrammarSlot
import org.jgll.grammar.slot.KeywordGrammarSlot
import org.jgll.grammar.slot.LastGrammarSlot
import org.jgll.grammar.slot.NonterminalGrammarSlot
import org.jgll.grammar.slot.TerminalGrammarSlot
import org.jgll.util.hashing.HashFunctionBuilder
import org.jgll.util.hashing.hashfunction.MurmurHash3
import scala.reflect.{BeanProperty, BooleanBeanProperty}
import collection.mutable._

//remove if not needed
import scala.collection.JavaConversions._

@SerialVersionUID(1L)
class Alternate(@BeanProperty val firstSlot: BodyGrammarSlot) extends Serializable {

  private val symbols = ListBuffer[BodyGrammarSlot]()

  @BeanProperty
  var condition: BodyGrammarSlot = _

  if (firstSlot == null) {
    throw new IllegalArgumentException("firstSlot cannot be null.")
  }

  var current = firstSlot

  if (firstSlot.isInstanceOf[LastGrammarSlot]) {
    symbols += firstSlot
  }

  while (!(current.isInstanceOf[LastGrammarSlot])) {
    symbols += current
    current = current.next()
  }

  def getSymbolAt(index: Int): Symbol = symbols.get(index).getSymbol

  def isEmpty(): Boolean = {
    firstSlot.isInstanceOf[EpsilonGrammarSlot]
  }

  def isNullable(): Boolean = {
    if (isEmpty) return true
    var slot = firstSlot
    while (!(slot.isInstanceOf[LastGrammarSlot])) {
      if (slot.isInstanceOf[TerminalGrammarSlot] || slot.isInstanceOf[KeywordGrammarSlot]) return false
      if (slot.isInstanceOf[NonterminalGrammarSlot] && !slot.isNullable) return false
      slot = slot.next()
    }
    true
  }

  def getSlotAt(index: Int): BodyGrammarSlot = symbols.get(index)

  def getLastSlot(): BodyGrammarSlot = symbols.get(symbols.size - 1)

  def size(): Int = symbols.size

  def getNonterminalAt(index: Int): HeadGrammarSlot = {
    val bodyGrammarSlot = symbols.get(index)
    if (!(bodyGrammarSlot.isInstanceOf[NonterminalGrammarSlot])) {
      throw new RuntimeException("The symbol at " + index + " should be a nonterminal.")
    }
    bodyGrammarSlot.asInstanceOf[NonterminalGrammarSlot]
      .getNonterminal
  }

  def setNonterminalAt(index: Int, head: HeadGrammarSlot) {
    val bodyGrammarSlot = symbols.get(index)
    if (!(bodyGrammarSlot.isInstanceOf[NonterminalGrammarSlot])) {
      throw new RuntimeException("The symbol at " + index + " should be a nonterminal.")
    }
    bodyGrammarSlot.asInstanceOf[NonterminalGrammarSlot]
      .setNonterminal(head)
  }

  override def toString(): String = {
    val sb = new StringBuilder()
    for (s <- symbols) {
      sb.append(s.getSymbol).append(" ")
    }
    sb.toString
  }

  def isBinary(head: HeadGrammarSlot): Boolean = {
    if (!(symbols.get(0).isInstanceOf[NonterminalGrammarSlot] && 
      symbols.get(symbols.size - 1).isInstanceOf[NonterminalGrammarSlot])) {
      return false
    }
    val firstNonterminal = symbols.get(0).asInstanceOf[NonterminalGrammarSlot]
    val lastNonterminal = symbols.get(symbols.size - 1).asInstanceOf[NonterminalGrammarSlot]
    head.getNonterminal == firstNonterminal.getNonterminal.getNonterminal && 
      head.getNonterminal == lastNonterminal.getNonterminal.getNonterminal
  }

  def isUnaryPrefix(head: HeadGrammarSlot): Boolean = {
    if (isBinary(head)) {
      return false
    }
    val index = symbols.size - 1
    if (!(symbols.get(index).isInstanceOf[NonterminalGrammarSlot])) {
      return false
    }
    val firstNonterminal = symbols.get(index).asInstanceOf[NonterminalGrammarSlot]
    head.getNonterminal == firstNonterminal.getNonterminal.getNonterminal
  }

  def isUnaryPostfix(head: HeadGrammarSlot): Boolean = {
    if (isBinary(head)) {
      return false
    }
    val index = 0
    if (!(symbols.get(index).isInstanceOf[NonterminalGrammarSlot])) {
      return false
    }
    val lastNonterminal = symbols.get(index).asInstanceOf[NonterminalGrammarSlot]
    head.getNonterminal == lastNonterminal.getNonterminal.getNonterminal
  }

  def `match`(list: ListBuffer[Symbol]): Boolean = {
    if (list.size != symbols.size) {
      return false
    }
    for (i <- 0 until symbols.size if symbols.get(i).getSymbol != list.get(i)) {
      return false
    }
    true
  }

  def `match`(set: Set[ListBuffer[Symbol]]): Boolean = {
    set.find(`match`(_)).isDefined
  }

  override def hashCode(): Int = {
    val hashBuilder = new HashFunctionBuilder(new MurmurHash3())
    var current = firstSlot
    while (current != null) {
      if (current.isInstanceOf[NonterminalGrammarSlot]) {
        hashBuilder.addInt(current.asInstanceOf[NonterminalGrammarSlot].getNonterminal
          .getNonterminal
          .hashCode)
      } else if (current.isInstanceOf[TerminalGrammarSlot]) {
        hashBuilder.addInt(current.asInstanceOf[TerminalGrammarSlot].getTerminal
          .hashCode)
      } else if (current.isInstanceOf[KeywordGrammarSlot]) {
        hashBuilder.addInt(current.asInstanceOf[KeywordGrammarSlot].getKeyword
          .hashCode)
      } else {
        hashBuilder.addInt(0)
      }
      current = current.next()
    }
    hashBuilder.hash()
  }

  override def equals(obj: Any): Boolean = {
    if (this == obj) {
      return true
    }
    if (!(obj.isInstanceOf[Alternate])) {
      return false
    }
    val other = obj.asInstanceOf[Alternate]
    if (this.size != other.size) {
      return false
    }
    var thisSlot = firstSlot
    var otherSlot = other.firstSlot
    while (!(thisSlot.isInstanceOf[LastGrammarSlot])) {
      if (!thisSlot.isNameEqual(otherSlot)) {
        return false
      }
      thisSlot = thisSlot.next()
      otherSlot = otherSlot.next()
    }
    true
  }

  def getSymbols(): Iterable[Symbol] = {
    new Iterable[Symbol]() {

      override def iterator(): Iterator[Symbol] = {
        new Iterator[Symbol]() {

          private var current: BodyGrammarSlot = firstSlot

          override def hasNext(): Boolean = current.next() != null

          override def next(): Symbol = {
            val s = current.getSymbol
            current = current.next()
            s
          }

          def remove() {
            throw new UnsupportedOperationException()
          }
        }
      }
    }
  }
}
